File: jungsol/SearchInRotatedSortedArray.java
public class SearchInRotatedSortedArray {
    public int search(int[] nums, int target) {
    	if(nums.length == 0) return -1;
    	int lo = 0, hi = nums.length-1;
    	while(lo < hi) {
    		int mid = (hi-lo) / 2 + lo;
    		if(nums[lo] <= nums[mid]) {
    			if(nums[lo] <= target && nums[mid] >= target) {
    				hi = mid;
    			} else {
    				lo = mid+1;
    			}
    		} else {
    			if(nums[mid] < target && nums[lo] > target) {
    				lo = mid+1;
    			} else {
    				hi = mid;
    			}
    		}
    	}
    	return (nums[lo] == target) ? lo : -1;
    }
}

File: jhkim/Search in_Rotated_Sorted_Array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0;
        int end = nums.size()-1;
        int mid = 0;
                
        while(true) {
            if(end<start)
                return -1;
            
            mid = (end+start)/2;
            
            if(nums[mid]==target)
                return mid;
            else if(nums[start]<nums[mid]) {
                if(start!=mid && nums[start]<=target && nums[mid-1]>=target)
                    end = mid-1;
                else
                    start = mid+1;
            }
            else {
                if(mid!=end && nums[mid+1]<=target && nums[end]>=target)
                    start = mid+1;
                else
                    end = mid-1;
            }
        }
    }
};
